[ENH] BEP 16: Diffusion modeling derivatives

arokem opened the pull request:
This transfers work that so far has been done in a separate repository: https://github.com/bids-standard/bids-bep016. 

I believe this is not too far from ready to be reviewed by the community, but also seeking input from @Lestropie, @francopestilli, @oesteban, @PeerHerholz.

Lestropie committed:
First commit for restructuring of DWI derivatives

Lestropie committed:
BEP016: Small fixes, and try to get links to headers working

Lestropie committed:
BEP016: More updates of internal header links

effigies committed:
ENH: Restore diffusion derivatives

Lestropie committed:
BEP016: Minor tweaks from first feedback round

- More clarity of distinction between requisite and optional files in output directory.
- Try using 3 spaces rather than 4 in non-code indentation; partly to try to get tables within dot point lists to render correctly, partly to improve editor software syntax highlighting.

Lestropie committed:
BEP016: Try to get links working within tables

Lestropie committed:
BEP016: More minor tweaking

- Various small re-wordings.
- Slightly more use of hyperlinks.
- Short introductions to "parameter terminology" and "data representations" sections.
- Be more explicit about normalised vs. non-normalised 3-vectors, so that structure more clusely mimics that of description of spherical coordinate representation.
- Rename hyperlink names to "parameter terminology" section to better separate from later "intrinsic / extrinsic model parameters" sections.

Lestropie committed:
BEP016: Transpose SH volume count tables

Lestropie committed:
BEP016: Provide example for parameter definitions

Lestropie committed:
BEP016: Change to single-file diffusion models

Based on suggestion in #5. If all model intrinsic parameters are incorporated into a single file, rather than dropping the "_parameter-<param>" field, instead enforce that parameter name "all" be used.

Lestropie committed:
BEP016: Initial explanation of filename

When introducing the file naming convention, give an example of the "_<model>" field.

Lestropie committed:
BEP016: File naming clarification

Re-arranged descriptions of intrinsic and extrinsic model parameters within the file naming section, and corrected a discordance in one dot point that was using an intrinsic model parameter filename path but discussing extrinsic model parameters.

Lestropie committed:
BEP016: Spelling fix

Lestropie committed:
BEP016: Fix missing anchor links

Lestropie committed:
BEP016: Move orientation specification

Provide information on specification of orientation data after the various models and model parameters have been explained.

oesteban committed:
MAINT: Split out tractography

This PR splits the tractography section from the diffusion derivatives
document, so that #5 is easier to merge.
The new ``05-diffusion-derivatives-tractography.md`` file will remain
orphaned, but kept there as a base for the time we tackle tractography.
It shouldn't be merged into the derivatives branch until it is ready.

arokem committed:
Merge pull request #9 from bids-standard/maint/split-tractography

[FIX] Split out tractography

oesteban committed:
Merge pull request #5 from bids-standard/dwi_derivatives_restructure

[ENH] Restructuring of DWI derivatives

effigies committed:
Merge remote-tracking branch 'upstream/master' into bep-016

Lestropie committed:
Merge remote-tracking branch 'bids-specification/master' into bep-016

Lestropie committed:
BEP016: Revise SH description

Revised based on MRtrix3/mrtrix3#1635.
Manual merge / cherry-pick of: c6539851, 8b278282, c6539851, ebbf0d3d, d7b4731d, 99e92e0a, 6bfc7849 due to unresolvable conflict in bids-standard/bids-bep016#8.

Lestropie committed:
BEP016: Initial conformity to markdown formatting

Lestropie committed:
BEP016: More formatting conformity changes

Lestropie committed:
BEP016: More formatting conformity changes

Lestropie committed:
BEP016: Attempted fixes to table cell padding

Lestropie committed:
BEP016: Minor fix to tractography derivatives for CI pass

Lestropie committed:
Merge pull request #30 from bids-standard/remark_fixes

[FIX] Formatting conformity to pass CI

Lestropie committed:
Merge branch 'bep-016' into sh_description_cherrypick

Conflicts:
	src/05-derivatives/05-diffusion-derivatives.md

Lestropie committed:
BEP016: Modify nomenclature around MRtrix3 SH coefficients

Lestropie committed:
BEP016: Revise preprocessed DWI data example

- Provide basic instructions rather than elaborating on rare use cases.
- Remove JSON dictionary on preprocessing steps utilised as these relate to provenance and are therefore out of scope.
Proposal for addressing bids-standard/bids-bep016#25.

Lestropie committed:
BEP016: Remove SH Descoteaux placeholder

Lestropie committed:
BEP016: Remove majority of models

Introduction and review of most models are intended to be deferred as per bids-specification/bids-bep016#39.

Lestropie committed:
BEP016: Formatting fix following removal of models

arokem committed:
Merge pull request #40 from bids-standard/remove_models

BEP016: Remove majority of models

francopestilli committed:
Merge pull request #37 from bids-standard/revise_preprocessed

[FIX] Revise preprocessed DWI data example

francopestilli committed:
Merge pull request #28 from bids-standard/sh_description_cherrypick

[ENH] SH description (take 2)

Lestropie committed:
BEP016: Strip out tractography content

arokem committed:
Fixes a typo caught by codespell

Lestropie committed:
Merge pull request #45 from bids-standard/typo-fix

FIX: typo caught by codespell

Lestropie committed:
BEP016: Initial draft of common filename suffix "model"

Lestropie committed:
BEP016: Rename "dti" to "tensor"

arokem committed:
Merge pull request #42 from bids-standard/strip_tractography

Strip out tractography content

Lestropie committed:
Merge pull request #47 from bids-standard/rename_tensor

BEP016: Rename "dti" to "tensor"

Lestropie committed:
Rename "_parameter-" entity to "_param-"

Lestropie committed:
DWI derivatives: Compulsory "parameter" entity

While the specification states that the "parameter" entity must always be defined---even if all model parameters are encoded within a single image, in which case value "all" must be used---two of the demonstrative examples did not obey such.

Lestropie committed:
Merge pull request #51 from bids-standard/param_entity

Rename "_parameter-" entity to "_param-"

Lestropie committed:
Merge remote-tracking branch 'origin/bep-016' into compulsory_parameter_entity

Lestropie committed:
DWI derivatives: Resolution between content changes

Resolves bids-standard/bids-bep016#52 (making "parameter" entity compulsory) against bids-standard/bids-bep016#51 (changing key for that parameter from "parameter" to "param").

Lestropie committed:
Merge remote-tracking branch 'origin/bep-016' into model_suffix

Conflicts:
	src/05-derivatives/05-diffusion-derivatives.md

Lestropie committed:
DWI derivatives: Fix tensor model example

Lestropie committed:
DWI derivatives: "model" as entity

francopestilli committed:
Merge pull request #46 from bids-standard/model_suffix

BEP016: Initial draft of common filename suffix "model"

Lestropie committed:
Merge branch 'bep-016' into compulsory_parameter_entity

Conflicts:
	src/05-derivatives/05-diffusion-derivatives.md

Lestropie committed:
Merge pull request #52 from bids-standard/compulsory_parameter_entity

DWI derivatives: Compulsory "parameter" entity

Lestropie committed:
Merge branch 'bep-016' into model_entity

Conflicts:
	src/05-derivatives/05-diffusion-derivatives.md

Lestropie committed:
Merge branch 'master' into bep-016

Lestropie committed:
Merge pull request #57 from bids-standard/model_entity

"model" as an entity

Lestropie committed:
DWI models: Restore renaming "dti" to "tensor"

Changes in #47 were lost in the process of conflict merging.

Lestropie committed:
Merge pull request #60 from bids-standard/restore_rename_tensor

DWI models: Restore renaming "dti" to "tensor"

Lestropie committed:
BEP016: Remove reference to old terminology

Formerly known as "intrinsic" model parameters are now (currently) referred to as simply "model parameters".

Lestropie committed:
Merge pull request #63 from bids-standard/revert_intrinsic

BEP016: Remove reference to old terminology

Lestropie committed:
BEP016: Revert "_param-all"

In cases where all model parameters can be encapsulated in a single NIfTI image, revert back to the case where the "param" entity is omitted from the file name.
Reverts https://github.com/bids-standard/bids-bep016/pull/52.

Lestropie committed:
Merge pull request #64 from bids-standard/no_paramall

Revert "_param-all"

Lestropie committed:
BEP016: Adopt "model fit parameters" terminology

In order to better disambiguate the various types of "parameters" defined, change "model parameters" to "model fit parameters".

Lestropie committed:
BEP016: Remove references to removed "pdf" data representation

Lestropie committed:
Merge pull request #65 from bids-standard/model_fit_parameters

BEP016: Adopt "model fit parameters" terminology

Lestropie committed:
Merge pull request #66 from bids-standard/remove_pdf

BEP016: Remove references to removed "pdf" data representation

Lestropie committed:
DWI derivatives: Make tensor a data representation

Lestropie committed:
DWI derivatives: Fix ModelDescription in exemplars

Specification requests field "ModelDescription", but exemplars at end of document instead erroneously used key "Model".

Lestropie committed:
Merge pull request #75 from bids-standard/model_description

DWI derivatives: Fix ModelDescription in exemplars

PeerHerholz committed:
update based on current BIDS spec master

This commit/PR updates the bids-BEP016 repository to incorporate the changes/current state in the bids-specification repo.

arokem committed:
Merge pull request #78 from PeerHerholz/update_specification

synch bids-bep016 with current state of bids-specification [MISC]

arokem committed:
Merge branch 'master' into bep-016

arokem committed:
Merge branch 'master' into bep-016

PeerHerholz committed:
Merge branch 'master' into bep-016

arokem committed:
Adds the descoteaux 2007 convention.

Based on https://github.com/dipy/dipy/blob/master/dipy/reconst/shm.py#L427-L430

arokem committed:
Merge pull request #80 from arokem/adds_descoteaux

Adds the descoteaux 2007 convention.

arokem committed:
Change from strange html encoding to "<" and ">"

arokem committed:
Merge pull request #81 from arokem/adds_descoteaux

Change from strange html encoding to "<" and ">"

arokem committed:
Fix codespell-detected British spellings.

arokem committed:
Merge pull request #82 from arokem/fix-british-spellings

Fix codespell-detected British spellings.

arokem committed:
Merge branch 'master' into bep-016

arokem committed:
Replace the latin "etc" with ellipses.

arokem committed:
Merge pull request #83 from arokem/fix-validation

Replace the latin "etc" with ellipses.

dependabot[bot] committed:
Bump actions/checkout from 3 to 4

Bumps [actions/checkout](https://github.com/actions/checkout) from 3 to 4.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v3...v4)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

arokem committed:
Merge branch 'master' into bep-016

Lestropie committed:
Merge branch 'bep-016' into tensor_as_representation

Lestropie committed:
Merge pull request #74 from bids-standard/tensor_as_representation

DWI derivatives: Make tensor a data representation

Lestropie committed:
BEP016: Major changes

- Reject suffices "mfp" (or earlier "_model") / "mfp" in favour of "_dwimap". Instead adopt distinction of "model metadata" vs. "parameter metadata".
- Reject use of advanced inheritance in favour of listing all relevant metadata for each data file in the sidecar JSON.
- Greater use of sub-dictionaries in JSON files to assist in separating metadata relevant to a model as a whole vs. only that particular parameter of the model.

Lestropie committed:
BEP016: Fis JSON formatting in demonstrative examples

Lestropie committed:
Diffusion models: Forbid negative spherical coordinate radius

arokem committed:
Diffusion model derivatives: Spelling fixes

arokem committed:
Merge pull request #92 from bids-standard/dwimap_and_no_inheritance

Major refactoring for upstream changes / blocks

arokem committed:
Merge branch 'master' into bep-016

arokem committed:
Apply US spelling. Addresses current codespell CI failure.

arokem committed:
Markdown table linting

arokem committed:
Get rid of a few latin phrases (i.e., "e.g.").

Lestropie committed:
DWI models: Define "bvec" orientation reference

Lestropie committed:
DWI models: Force presence of "param" entity

Lestropie committed:
DWI models: add "ParameterURL" metadata field

Lestropie committed:
Merge remote-tracking branch 'origin/bep-016' into dependabot/github_actions/actions/checkout-4

Conflicts:
	.github/workflows/schemacode_ci.yml
	.github/workflows/validation.yml

Lestropie committed:
Clarify metadata fields relevance (#102)

* DWI models: Clarify metadata fields relevance

* DWI models: Typo fix

---------

Co-authored-by: Ariel Rokem <arokem@gmail.com>

Lestropie committed:
Merge pull request #87 from bids-standard/dependabot/github_actions/actions/checkout-4

Bump actions/checkout from 3 to 4

Lestropie committed:
Merge pull request #97 from arokem/us-spelling

Apply US spelling. Addresses current codespell CI failure.

arokem committed:
Merge branch 'bep-016' into metadata_parameterurl

arokem committed:
Merge pull request #101 from bids-standard/metadata_parameterurl

Add "ParameterURL" metadata field

arokem committed:
Merge branch 'bep-016' into force_param_entity

arokem committed:
Merge pull request #99 from bids-standard/force_param_entity

DWI models: Force presence of "param" entity

arokem committed:
Removes one more i.e.

Lestropie committed:
Merge pull request #103 from arokem/one_more_latin

Removes one more i.e.

Lestropie committed:
BEP016: Utilize filesystem macro

Lestropie committed:
BEP016: Reformat demonstrative examples

Lestropie committed:
BEP016: Don't use filesystem macros for templates

Lestropie committed:
BEP016: Comment filesystem macros

Lestropie committed:
Merge pull request #105 from bids-standard/filesystem_macros

Filesystem macros

Lestropie committed:
DWI models: Update bvec reference

- FSL bedpostx command fibre orientations stored as spherical coordinates are confirmed to use the "bvec" reference frame.
- Enforcing azimuth angle to obey the right hand rule about the zenith axis would necessitate direct modification of image intensities from bedpostx outputs to store. Therefore it would be preferable to instead define the sign of the azimuth angle based on the direction of the second reference axis.
Closes #95.
Closes #94.

arokem committed:
Merge branch 'master' into bep-016

Lestropie committed:
Merge remote-tracking branch 'origin/bep-016' into bvec_reference_axes

Conflicts:
	src/derivatives/05-diffusion-derivatives.md

Lestropie committed:
Merge pull request #98 from bids-standard/bvec_reference_axes

DWI models: Define "bvec" orientation reference

arokem committed:
Merge branch 'master' into bep-016

arokem committed:
Merge branch 'master' into bep-016

arokem commented:
Forgot to tag @mattcieslak, whose input here also useful. And probably others, please tag and add to this discussion.

PeerHerholz commented:
Hey @arokem and everyone,

thx @arokem! Yeah, I think once it's decided that the current `markdown` and respective information/proposals are in a good status we would create a couple of example datasets (I can update the [conversion software](https://github.com/PeerHerholz/bids_bep16_conv) respectively). I do agree that the examples should be relatively easy, especially considering that we would provide empty imaging files and focus on the `metadata`. The `schema` appears to be a different beast and having someone from the maintainers/developers involved would be great.

arokem committed:
Fixes links.

arokem committed:
Use naming convention without number prefix.

arokem committed:
Use US spelling of "realizations".

arokem committed:
Allow "burnin", which is a technical term used in describing ball and stick.

arokem committed:
Try to sort out markdown linting.

effigies commented:
> One general question: does the validator currently look into the content of data files to validate that their content is sensible?

It validates tsv contents and NIfTI headers.

Validating NIfTI data blocks is likely to be significantly slower, and would probably make sense for a separate tool like [miniQC](https://github.com/openneuroorg/miniqc).

What kind of check are you thinking?

arokem commented:
One thing that comes to mind is to validate the shape of the data. For example, the tensor should have six elements on the last dimension; CSD coefficients would also have a certain dimensionality that can be checked. This can be done without reading the data. So, if there is precedent for this kind of check, that's good.

effigies commented:
Yes, checking NIfTI shapes is doable:

https://github.com/bids-standard/bids-specification/blob/b9a66732e67de7bc51402b7ed0a73a1e58ce85d6/src/schema/rules/checks/func.yaml#L57-L67

You'll want to use `make_sidecar_table` if it corresponds to a sidecar schema rule, but for describing deeper models, you'll probably want `make_metadata_table`, e.g.,

https://github.com/bids-standard/bids-specification/blob/b9a66732e67de7bc51402b7ed0a73a1e58ce85d6/src/common-principles.md?plain=1#L539-L561

arokem committed:
Merge branch 'master' into bep-016

arokem committed:
Remove remaining instances of `_model` suffix.

arokem committed:
Adds macros for tables. Adds an example of NODDI fit.

arokem committed:
Start converting metadata tables to macros.

arokem committed:
Fixes malformed json in table.

arokem committed:
More malformed json fixed.

arokem committed:
More metadata tables converted with macro.

arokem committed:
Fixes malformed json in macro.

arokem committed:
Removes backticks in json.

arokem committed:
Remove side_car_table call that makes no sense here.

arokem committed:
Adds diffusion derivatives to toc.

arokem committed:
Fix bval/bvec extensions.

arokem committed:
Adding parameters needed to describe a noddi fit.

arokem committed:
diso and dpar should be in the Parameters dict of the model.

arokem committed:
Fixes key name in model metadata dict.

arokem committed:
Use mm-related values, because b-values are already encoded in that scale.

This diverges from the usual recommendation to use SI units, but makes
sense in this field, where values of mm^2/s are used to describe diffusivity
in many different places.

tsalo committed:
Merge branch 'master' into bep-016

arokem committed:
Adds parameterURL for noddi parameters.

tsalo commented on src/derivatives/diffusion-derivatives.md:
```diff
@@ -0,0 +1,1125 @@
+   }
+) }}
+
+Dictionary `"Model["Parameters"]"` has the following reserved keywords that may be applicable to a broad range of models:
```
Pinging @arokem @effigies

tsalo replied:
The only thing I can think of is the provenance BEP, but that uses json.ld files instead of BIDS sidecars.

tsalo committed:
Add metadata fields to schema.

I didn't use any references so the glossary links will be broken.

tsalo committed:
Address remark warnings.

arokem committed:
Merge pull request #8 from tsalo/bep-016-tsalo

Add metadata fields to schema

tsalo committed:
Try to fix warnings.

arokem committed:
Merge pull request #9 from tsalo/bep-016-tsalo-02

Fix yamllint and remark warnings

arokem committed:
Move comments into GitHub PR discussion.

arokem committed:
Merge branch 'bep-016' of github.com:arokem/bids-specification into bep-016

mattcieslak commented on src/derivatives/diffusion-derivatives.md:
```diff
@@ -0,0 +1,1025 @@
+
+        | ***l<sub>max</sub>*** | 0 | 2 | 4 | 6 | 8 | 10 | ... |
+        | --------------------- | --:| --:| --: | --: | --: | --: | :--: |
+        | ***N***               | 1 | 6 | 15 | 28 | 45 | 66 | ... |
```
I'd like to be able to represent asymmetric odfs in BIDS
